File: apm/src/main/java/io/github/yemyatthu1990/apm/collector/MetricsCollector.java
package io.github.yemyatthu1990.apm.collector;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.Collection;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;
import java.util.function.BiConsumer;

class MetricsCollector implements Mapper {
    private final ConcurrentMap<String, String> cache;

    public MetricsCollector() {
        cache = new ConcurrentHashMap<>();
    }

    @Override
    public int size() {
        return cache.size();
    }

    @Override
    public boolean isEmpty() {
        return cache.isEmpty();
    }

    @Override
    public boolean hasKey(@NonNull String key) {
        return cache.containsKey(key);
    }

    @Nullable
    @Override
    public String getValue(@NonNull String key) {
        return cache.get(key);
    }

    @Nullable
    @Override
    public ConcurrentMap<String, String> map() {
        return cache;
    }

    @Nullable
    @Override
    public Set<String> getKeySet() {
        return cache.keySet();
    }

    @Nullable
    @Override
    public Collection<String> valueSet() {
        return cache.values();
    }

    @Override
    public void put(@NonNull String key, @NonNull String value) {
        if (!value.isEmpty()) {
            cache.put(key, value);
        }
    }

    @Override
    public void putAll(@Nullable ConcurrentMap<String, String> map) {
        if (map != null) {
            for (String key: map.keySet()) {
                String value = map.get(key);
                if (value != null && !value.isEmpty()) {
                    cache.put(key, value);
                }
            }

        }
    }

    @Override
    public void clear() {
        cache.clear();
    }
}

File: apm/src/main/java/io/github/yemyatthu1990/apm/collectors/MetricsCollector.kt
package io.github.yemyatthu1990.apm.collectors

import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ConcurrentMap

open class MetricsCollector : Mapper{
    private val cache: ConcurrentMap<String, String>
    init {
        cache = ConcurrentHashMap()
    }
    override fun size() = cache.size

    override val isEmpty = cache.isEmpty()
    override fun hasKey(key: String) = cache.containsKey(key)

    override fun getValue(key: String): String = cache.getValue(key)


    override fun map(): ConcurrentMap<String, String> = cache

    override val keySet: Set<String> = cache.keys

    override fun valueSet(): Collection<String> = cache.values
    override fun put(key: String, value: String) {
        cache[key] = value
    }

    override fun putAll(map: ConcurrentMap<String, String>?) {
        map?.let {
            cache.putAll(map)
        }
    }


    override fun clear() {
        cache.clear()
    }


}